### ISSUE TITLE
Accessing token vector attribute hangs randomly for some texts

### ISSUE THREAD

**Issue by buhrmann** [explosion/spaCy#7915]
## How to reproduce the behaviour

The following code will hang at a different token when accessing its vector each time it's executed. This means it will also fail when accessing the averaged document vector.

``` python
import spacy

en = spacy.load('en_core_web_md')
print("Model version:", en.meta["version"])

txt = """..."""
print("Text #nchars:", len(txt))

doc = en(txt)
print("Doc #tokens:", len(doc))

for i, t in enumerate(doc):
    print(i, t.has_vector, len(t.vector))
```

I've shared the (a?) text this fails on in [this gist](https://gist.github.com/buhrmann/33aaf371ae4819a3505ba38aa8d1a922). It is arguable a rather odd text (a not exactly clean scrape from a public web page including a long list of IDs). Although I wouldn't be surprised to get non-sensical results in this case, I don't know how to predict what sort of text will fail in this manner, and so it's difficult to mitigate if the source of texts is not known.

The output is

```
Model version: 3.0.0
Text #nchars: 150000
Doc #tokens: 5644
0 True 300
1 False 300
2 True 300
3 True 300
4 False 300
5 True 300
6 True 300
7 True 300
8 True 300
9 True 300
10 True 300
... hangs here randomly after a couple thousand tokens
```

## Your Environment

- **spaCy version:** 3.0.6
- **Platform:** Linux-5.6.0-1055-oem-x86_64-with-glibc2.10
- **Python version:** 3.8.8
- **Pipelines:** en_core_web_sm (3.0.0), en_core_web_md (3.0.0)

**Comment by buhrmann** [comment]
Here is some more, hopefully useful, information. I've compared execution times for the first N characters in a normal text vs. the weird text posted above:

```
N Chars    Bad text  Normal text
---------------------------------
40,000     54.12s    1.18s
35,000     25.80s    0.84s
30,000      9.80s    0.69s
25,000      1.78s    0.58s
20,000      0.51s    0.45s
```

No idea still what it is about that text exactly that makes spaCy behave this way.

**Comment by adrianeboyd** [comment]
Thanks for the report, this does look like a bug. I don't think it's random, though, it seems to happen for a token that's 100K+ characters long.

There's a kind of fasttext-y ngram backoff that should be skipped unless `Vocab.get_vector(minn)` is set, and it looks like it's not skipping it correctly, so whatever it's doing (unnecessarily) for 100K characters is slow. (My opinion is that this should be removed and you can use a custom vector hook instead, since it doesn't really work as intended in practice from a typical vectors table. But this bug is at least easy to fix even if we keep it.)

**Comment by buhrmann** [comment]
Thanks for the quick response. In my updated/edited comment above I've identified the long list of ID-like strings as the culprit. Perhaps that throws the tokenization off and generates very long "words"?

**Comment by buhrmann** [comment]
Could you perhaps point me at the code you mention that should be skipped? That way I could write my own hook as you suggest, but without the offending bit

**Comment by adrianeboyd** [comment]
Sure, see the linked issue above. You can access `doc.vocab.vectors` or `token.vocab.vectors` to get the vectors in the user hook.

**Comment by buhrmann** [comment]
Thanks! So basically I'd have to avoid `vocab.get_vector` and use something like this, right?

```
def vector(token):
    vocab = token.vocab
    if token.orth in vocab.vectors.key2row:
        return vocab.vectors[token.orth]
    return np.zeros((vocab.vectors_length,), dtype="f")
```
And once your fix is merged, not setting minn, maxnn will have the same effect (of not running the fasttext-y code path and directly returning the 0-vector).

**Comment by adrianeboyd** [comment]
If you're potentially loading the model on GPU, you'll get mixed `cupy.ndarray` and `numpy.ndarray` results from this, but as long as that doesn't cause issues, I think that should fine. Not setting `minn` should make it skip the whole fasttext-y bit (which again, doesn't really work from a standard vector table that wasn't trained using subwords).

None of this affects the statistical models, which access the underlying vector table directly.

**Comment by buhrmann** [comment]
Ok, thanks. The workaround works fine for me, and since it's fixed in #7925 I think you can close this issue.
